loadtest/loadtest.go

```diff
@@ -75,26 +77,42 @@ func (lt *LoadTester) AddUsers(numUsers int) (int, error) {
 // addUser is an internal API called from Run and AddUsers both.
 // DO NOT call this by itself, because this method is not protected by a mutex.
 func (lt *LoadTester) addUser() error {
-	activeUsers := len(lt.controllers)
+	activeUsers := len(lt.activeControllers)
 	if activeUsers == lt.config.UsersConfiguration.MaxActiveUsers {
 		return ErrMaxUsersReached
 	}
-	userId := activeUsers + 1
+
+	idleUsers := len(lt.idleControllers)
+	userId := activeUsers + idleUsers + 1
 	// If specified by the config, we randomly pick an existing user again,
 	// to simulate multiple sessions.
 	if activeUsers != 0 && rand.Int()%lt.config.UsersConfiguration.AvgSessionsPerUser != 0 {
 		userId = rand.Intn(activeUsers)
 	}
-	controller, err := lt.newController(userId, lt.statusChan)
-	if err != nil {
-		return fmt.Errorf("could not create controller: %w", err)
+
+	var controller control.UserController
+
+	if idleUsers > 0 {
+		controller = lt.idleControllers[0]
+		lt.idleControllers = lt.idleControllers[1:]
 	}
+
+	if controller == nil {
```

Comment:
Again here I find the logic to be slightly misleading. Is it possible for `controller` to be `nil` if `idleUsers` is greater than zero?
It appears like this could be a simple if/else on `len(lt.idleControllers)`.